cylc/flow/templatevars.py

```diff
@@ -77,22 +83,104 @@ def eval_var(var):
+    in_quotes = False
+    buffer = ''
+    for char in stringlist:
+        if char in {'"', "'"}:
+            in_quotes = not in_quotes
+        elif not in_quotes and char == ',':
+            list_.append(buffer)
+            buffer = ''
+        else:
+            buffer += char
+    list_.append(buffer)
+    return list_
+
+
+def load_template_vars(
+    template_vars: Optional[List[str]] = None,
+    template_vars_file: Union[Path, str, None] = None,
+    templatevars_lists: Optional[List[str]] = None
+) -> Dict[str, Any]:
     """Load template variables from key=value strings."""
-    res = {}
+    keys: set = set()
+    invalid_vars: str = ''
+
+    # Parse Template vars set by file (-S)
+    file_tvars: Dict = {}
     if template_vars_file:
         with open(template_vars_file, 'r') as handle:
             for line in handle:
                 line = line.strip().split("#", 1)[0]
                 if not line:
                     continue
-                key, val = line.split("=", 1)
-                res[key.strip()] = eval_var(val.strip())
+                try:
+                    key, val = line.split("=", 1)
+                except ValueError:
+                    invalid_vars += f'\n * {line}'
+                    continue
+                file_tvars[key.strip()] = eval_var(val.strip())
+                keys.add(key.strip())
+
+    cli_tvars: Dict = {}
+    tvars_lists: Dict = {}
+    for input_, result, func in (
+        (template_vars, cli_tvars, eval_var),
+        (templatevars_lists, tvars_lists, parse_string_list)
+    ):
+        for pair in input_ or []:
+            # Handle ValueError
+            try:
+                key, val = pair.split("=", 1)
+            except ValueError:
+                invalid_vars += f'\n * {pair}'
+                continue
+            key, val = key.strip(), val.strip()
+            if key in result:
+                LOG.warning(OVERWRITE_WARNING.format(key))
+            result[key] = func(val)
+            keys.add(key)
+
+    # Raise an error if there are any args without the form key=value.:
+    if invalid_vars:
+        raise InputError(
+            'Template variables should be in the form key=value(s) -'
+            '\nThe following arguments did not validate:'
+            + invalid_vars
+        )
+
+    # Explicitly record which source of tvars over-rides which.
+    res = {}
+    badkeys = ''
+    for key in keys:
+        if key in cli_tvars and key in tvars_lists:
+            badkeys += (
+                f"\n * {key}={cli_tvars[key]} and {key}={tvars_lists[key]}")
+        else:
+            res[key] = cli_tvars.get(
+                key, tvars_lists.get(
+                    key, file_tvars.get(key)))
+
+    # Raise an error if there are any key-clashes between tvars and tvars_list
+    if badkeys:
+        raise InputError(
+            'Setting the same variable using -s and -z is not permitted -'
+            '\nThe following clashing template variables:\n * '
```

Comment:
```suggestion
            "You can't set the same template variable with both -s and -z:"
```
Statements like "not permitted" not really necessary; that's implied by the error message itself. And the second sentence kind of restates the first.